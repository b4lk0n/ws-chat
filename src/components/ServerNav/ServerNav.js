import React, { useState, useMemo } from 'react';
import { arrayOf, shape, oneOfType, number, string, instanceOf, bool, func } from 'prop-types';
import cx from 'classnames';
import { TextField } from '../TextField';
import { IconButton } from '../IconButton';
import { ServerNavItem } from '../ServerNavItem';

import styles from './ServerNav.module.css';

export const ServerNav = ({ servers, value, onChange, onCreate, className, ...props }) => {
  const [search, setSearch] = useState('');

  const filteredServers = useMemo(() => {
    return search ? servers.filter(({ url }) => url.includes(search.toLowerCase())) : servers;
  }, [servers, search]);

  return (
    <div data-testid={'server-nav'} {...props} className={cx(styles.root, className)}>
      <div className={styles.header}>
        <TextField
          placeholder={'Search'}
          value={search}
          onChange={setSearch}
          icon={'search'}
          withClear
          className={styles.searchField}
          data-testid={'server-nav-search-field'}
        />

        <IconButton
          icon={'add_circle_outline'}
          size={'large'}
          onClick={onCreate}
          data-testid={'server-nav-create-button'}
        />
      </div>

      <div className={styles.serversList}>
        {filteredServers.map(server => (
          <ServerNavItem
            server={server}
            key={server.id}
            onClick={onChange}
            className={cx(styles.serverItem, {
              [styles.active]: value === server,
            })}
          />
        ))}
      </div>
    </div>
  );
};

const serverShape = shape({
  id: oneOfType([string, number]).isRequired,
  url: string.isRequired,
  updatedAt: instanceOf(Date),
  connectionStatus: bool,
});
ServerNav.propTypes = {
  servers: arrayOf(serverShape).isRequired,
  onChange: func.isRequired,
  onCreate: func.isRequired,
  value: serverShape,
  className: string,
};
