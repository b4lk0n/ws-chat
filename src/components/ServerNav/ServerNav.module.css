.root {
  background-color: var(--surface);
}

.header {
  padding: 8px 8px 8px 16px;
  display: flex;
  align-items: center;
}

.searchField {
  flex: 1;
  margin-right: 4px;
}

.serversList {
  padding: 16px 0;
}

.serverItem {
  transition: background-color 0.1s ease-in-out;
}

.serverItem:focus,
.serverItem:hover {
  background-color: var(--outlineLight);
}

.serverItem.active {
  background-color: var(--outline);
  cursor: default;
  pointer-events: none;
}
