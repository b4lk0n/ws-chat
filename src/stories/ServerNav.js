import React, { useState } from 'react';
import { action } from '@storybook/addon-actions';
import { ServerNav } from '../components';

export default {
  title: 'ServerNav',
};

const servers = [
  { id: 1, url: 'wss://server1.com', updatedAt: new Date(1580475448818) },
  { id: 2, url: 'wss://server2.com', updatedAt: new Date(1580475471956), connectionStatus: true },
  { id: 3, url: 'wss://server3.com', updatedAt: new Date(1580475493072), connectionStatus: false },
  { id: 4, url: 'wss://server4.com' },
];

export const Default = () => {
  const [value, setValue] = useState();

  return (
    <div style={{ width: 320 }}>
      <ServerNav servers={servers} value={value} onChange={setValue} onCreate={action('onCreate')} />
    </div>
  );
};
