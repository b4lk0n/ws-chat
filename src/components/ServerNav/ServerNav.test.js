import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { ServerNav } from './ServerNav';

const servers = [
  { id: 1, url: 'wss://server1.com', updatedAt: new Date(1580475448818) },
  { id: 2, url: 'wss://server2.com', updatedAt: new Date(1580475471956), connectionStatus: true },
  { id: 3, url: 'wss://server3.com', updatedAt: new Date(1580475493072), connectionStatus: false },
  { id: 4, url: 'wss://server4.com' },
];

describe('<ServerNav />', () => {
  it('should renders correctly', () => {
    const changeHandler = jest.fn();
    const createHandler = jest.fn();
    const { getByTestId, getAllByTestId } = render(
      <ServerNav servers={servers} onChange={changeHandler} onCreate={createHandler} />
    );
    const nav = getByTestId('server-nav');

    expect(changeHandler).not.toBeCalled();
    expect(createHandler).not.toBeCalled();
    fireEvent.click(getByTestId('server-nav-create-button'));
    expect(createHandler).toBeCalledTimes(1);
    fireEvent.click(getAllByTestId('server-nav-item')[0]);
    expect(changeHandler).toBeCalledTimes(1);
    expect(changeHandler).toBeCalledWith(servers[0]);

    expect(nav).toMatchSnapshot();
  });
});
