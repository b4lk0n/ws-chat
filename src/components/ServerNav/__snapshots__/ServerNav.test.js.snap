// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ServerNav /> should renders correctly 1`] = `
<div
  class="root"
  data-testid="server-nav"
>
  <div
    class="header"
  >
    <div
      class="root withIcon withClear searchField"
      data-testid="textfield"
    >
      <input
        data-testid="textfield-input"
        placeholder="Search"
        value=""
      />
      <span
        class="root undefined material-icons icon"
        data-testid="icon"
      >
        search
      </span>
      
    </div>
    <button
      class="root large"
      data-testid="server-nav-create-button"
    >
      <span
        class="root undefined material-icons"
        data-testid="icon"
      >
        add_circle_outline
      </span>
    </button>
  </div>
  <div
    class="serversList"
  >
    <div
      class="root clickable serverItem"
      data-testid="server-nav-item"
    >
      <div
        class="info"
      >
        <span
          class="root undefined serverUrl"
          data-testid="text"
        >
          wss://server1.com
        </span>
        <span
          class="root caption updatedAt"
          data-testid="text"
        >
          31 Jan 2020, 14:57
        </span>
      </div>
      <div
        class="connectionStatus"
      >
        <span
          class="root undefined"
          data-testid="icon"
        >
          <span
            class="root"
            data-testid="connection-status-icon"
          />
        </span>
      </div>
    </div>
    <div
      class="root clickable serverItem"
      data-testid="server-nav-item"
    >
      <div
        class="info"
      >
        <span
          class="root undefined serverUrl"
          data-testid="text"
        >
          wss://server2.com
        </span>
        <span
          class="root caption updatedAt"
          data-testid="text"
        >
          31 Jan 2020, 14:57
        </span>
      </div>
      <div
        class="connectionStatus"
      >
        <span
          class="root undefined"
          data-testid="icon"
        >
          <span
            class="root success"
            data-testid="connection-status-icon"
          />
        </span>
      </div>
    </div>
    <div
      class="root clickable serverItem"
      data-testid="server-nav-item"
    >
      <div
        class="info"
      >
        <span
          class="root undefined serverUrl"
          data-testid="text"
        >
          wss://server3.com
        </span>
        <span
          class="root caption updatedAt"
          data-testid="text"
        >
          31 Jan 2020, 14:58
        </span>
      </div>
      <div
        class="connectionStatus"
      >
        <span
          class="root undefined"
          data-testid="icon"
        >
          <span
            class="root failure"
            data-testid="connection-status-icon"
          />
        </span>
      </div>
    </div>
    <div
      class="root clickable serverItem"
      data-testid="server-nav-item"
    >
      <div
        class="info"
      >
        <span
          class="root undefined serverUrl"
          data-testid="text"
        >
          wss://server4.com
        </span>
      </div>
      <div
        class="connectionStatus"
      >
        <span
          class="root undefined"
          data-testid="icon"
        >
          <span
            class="root"
            data-testid="connection-status-icon"
          />
        </span>
      </div>
    </div>
  </div>
</div>
`;
